File: Backend/src/logic/ai/syllabus_planner.rs
use crate::repository::Repositories;
use anyhow::{anyhow, Result};
use chrono::{Datelike, NaiveDate};
use serde_json::{json, Value};
use sqlx::Row;
use std::sync::Arc;

pub struct SyllabusPlanner {
    pub repos: Arc<Repositories>,
}

impl SyllabusPlanner {
    pub fn new(repos: Arc<Repositories>) -> Self {
        Self { repos }
    }

    /// Plot annual syllabus for a class-subject: distribute chapters across 4 quarters
    /// based on weightage, respecting academic year (April 1 – March 31).
    pub async fn annual_syllabus_plot(
        &self,
        school_id: &str,
        class_id: &str,
        subject_id: &str,
        academic_year: i32, // e.g. 2026 = Apr 2026 – Mar 2027
    ) -> Result<Value> {
        let mut conn = self.repos.db_client.acquire_tenant_connection(school_id).await?;

        // 1. Fetch all chapters for this subject
        let chapters = sqlx::query(
            "SELECT id, name, weightage, quarter FROM chapters WHERE school_id = $1 AND subject_id = $2 ORDER BY sequence_order"
        )
        .bind(school_id).bind(subject_id)
        .fetch_all(&mut *conn)
        .await?;

        if chapters.is_empty() {
            return Err(anyhow!("No chapters found for subject {}", subject_id));
        }

        // 2. Holiday lookup (all holidays in the academic year)
        let year_start = NaiveDate::from_ymd_opt(academic_year, 4, 1).unwrap();
        let year_end = NaiveDate::from_ymd_opt(academic_year + 1, 3, 31).unwrap();
        let holidays: Vec<NaiveDate> = sqlx::query_scalar(
            "SELECT date FROM holidays WHERE school_id = $1 AND date >= $2 AND date <= $3"
        )
        .bind(school_id).bind(year_start).bind(year_end)
        .fetch_all(&mut *conn)
        .await
        .unwrap_or_default();

        // 3. Quarter boundaries
        let quarters = vec![
            ("Q1", NaiveDate::from_ymd_opt(academic_year, 4, 1).unwrap(), NaiveDate::from_ymd_opt(academic_year, 6, 30).unwrap()),
            ("Q2", NaiveDate::from_ymd_opt(academic_year, 7, 1).unwrap(), NaiveDate::from_ymd_opt(academic_year, 9, 30).unwrap()),
            ("Q3", NaiveDate::from_ymd_opt(academic_year, 10, 1).unwrap(), NaiveDate::from_ymd_opt(academic_year, 12, 31).unwrap()),
            ("Q4", NaiveDate::from_ymd_opt(academic_year + 1, 1, 1).unwrap(), NaiveDate::from_ymd_opt(academic_year + 1, 3, 31).unwrap()),
        ];

        // 4. Compute available teaching days per quarter
        let mut quarter_days: Vec<(String, Vec<NaiveDate>)> = Vec::new();
        for (q_name, q_start, q_end) in &quarters {
            let mut days = Vec::new();
            let mut d = *q_start;
            while d <= *q_end {
                let dow = d.format("%u").to_string().parse::<u32>().unwrap_or(0);
                if dow != 7 && !holidays.contains(&d) {
                    days.push(d);
                }
                d += chrono::Duration::days(1);
            }
            quarter_days.push((q_name.to_string(), days));
        }

        // 5. Distribute chapters across quarters by weightage
        let total_weight: i32 = chapters.iter()
            .map(|r| r.get::<Option<i32>, &str>("weightage").unwrap_or(1))
            .sum();

        if total_weight == 0 {
            return Err(anyhow!("Total weightage is zero, cannot distribute"));
        }

        // Use the same optimal partition algorithm as get_auto_syllabus
        let n = chapters.len();
        let per_quarter_target = total_weight as f64 / 4.0;

        // Find optimal split points (i, j, k) for Q1, Q2, Q3, Q4
        let (p1, p2, p3) = if n < 4 {
            (0usize, 1usize.min(n.saturating_sub(1)), 2usize.min(n.saturating_sub(1)))
        } else {
            let mut best = (0usize, 1usize, 2usize);
            let mut min_err = f64::MAX;
            for i in 0..n {
                for j in (i + 1)..n {
                    for k in (j + 1)..n {
                        if k >= n - 1 { continue; }
                        let w1: i32 = chapters[0..=i].iter().map(|r| r.get::<Option<i32>, &str>("weightage").unwrap_or(1)).sum();
                        let w2: i32 = chapters[(i + 1)..=j].iter().map(|r| r.get::<Option<i32>, &str>("weightage").unwrap_or(1)).sum();
                        let w3: i32 = chapters[(j + 1)..=k].iter().map(|r| r.get::<Option<i32>, &str>("weightage").unwrap_or(1)).sum();
                        let w4: i32 = chapters[(k + 1)..].iter().map(|r| r.get::<Option<i32>, &str>("weightage").unwrap_or(1)).sum();
                        let err = (w1 as f64 - per_quarter_target).powi(2)
                            + (w2 as f64 - per_quarter_target).powi(2)
                            + (w3 as f64 - per_quarter_target).powi(2)
                            + (w4 as f64 - per_quarter_target).powi(2);
                        if err < min_err { min_err = err; best = (i, j, k); }
                    }
                }
            }
            best
        };

        // Assign chapters to quarters
        let quarter_chapter_ranges: Vec<(std::ops::RangeInclusive<usize>, &str)> = vec![
            (0..=p1, "Q1"),
            ((p1 + 1)..=p2, "Q2"),
            ((p2 + 1)..=p3, "Q3"),
            ((p3 + 1)..=(n.saturating_sub(1)), "Q4"),
        ];

        let mut inserted = Vec::new();

        for (range, quarter_name) in &quarter_chapter_ranges {
            let q_idx = match *quarter_name {
                "Q1" => 0usize, "Q2" => 1, "Q3" => 2, _ => 3,
            };
            let q_days = &quarter_days[q_idx].1;

            for (ch_idx_in_quarter, ch_idx) in range.clone().enumerate() {
                let row = &chapters[ch_idx];
                let chapter_id: i32 = row.get("id");
                let chapter_name: String = row.get("name");
                let ch_weight: i32 = row.get::<Option<i32>, &str>("weightage").unwrap_or(1);

                // Plan dates: distribute evenly within the quarter
                let range_size = range.clone().count().max(1);
                let day_start = (ch_idx_in_quarter * q_days.len().max(1)) / range_size;
                let day_end = ((ch_idx_in_quarter + 1) * q_days.len().max(1)) / range_size;
                let planned_start = q_days.get(day_start.min(q_days.len().max(1).saturating_sub(1))).copied();
                let planned_end = q_days.get(day_end.min(q_days.len().max(1).saturating_sub(1))).copied();

                let period_count = (ch_weight as f64 / total_weight as f64 * (q_days.len() as f64 * 6.0)).ceil() as i32;

                sqlx::query(
                    "INSERT INTO syllabus_calendar (school_id, class_id, subject_id, chapter_id, \
                     planned_start_date, planned_end_date, period_count, quarter, status) \
                     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, 'pending') \
                     ON CONFLICT (school_id, class_id, subject_id, chapter_id, quarter) \
                     DO UPDATE SET planned_start_date = EXCLUDED.planned_start_date, \
                                   planned_end_date = EXCLUDED.planned_end_date, \
                                   period_count = EXCLUDED.period_count"
                )
                .bind(school_id).bind(class_id).bind(subject_id).bind(chapter_id)
                .bind(&planned_end.unwrap_or(q_days.first().copied().unwrap_or(year_start)))
                .bind(&planned_end.unwrap_or(q_days.last().copied().unwrap_or(year_end)))
                .bind(period_count)
                .bind(quarter_name)
                .execute(&mut *conn)
                .await?;

                inserted.push(json!({
                    "chapterId": chapter_id,
                    "chapterName": chapter_name,
                    "quarter": quarter_name,
                    "plannedStartDate": planned_start.map(|d| d.to_string()),
                    "plannedEndDate": planned_end.map(|d| d.to_string()),
                    "periodCount": period_count,
                }));
            }
        }

        Ok(json!({
            "success": true,
            "academicYear": format!("{}-{}", academic_year, academic_year + 1),
            "subjectId": subject_id,
            "totalChapters": chapters.len(),
            "syllabus": inserted,
        }))
    }

    /// Micro-plan period-level assignments for a date range.
    /// For each day with timetable slots for this class+subject, assign specific topics.
    pub async fn micro_plan_period_level(
        &self,
        school_id: &str,
        class_id: &str,
        subject_id: &str,
        from_date: &str,
        to_date: &str,
    ) -> Result<Value> {
        let mut conn = self.repos.db_client.acquire_tenant_connection(school_id).await?;

        let from = NaiveDate::parse_from_str(from_date, "%Y-%m-%d")?;
        let to = NaiveDate::parse_from_str(to_date, "%Y-%m-%d")?;

        // 1. Fetch timetable slots for this class+subject in date range
        let slots = sqlx::query(
            "SELECT ts.day_of_week, ts.period_number, ts.teacher_id, tc.config_id \
             FROM timetable_slots ts \
             JOIN timetable_configs tc ON tc.config_id = ts.config_id AND tc.school_id = ts.school_id \
             WHERE ts.school_id = $1 AND ts.class_id = $2 AND ts.subject_id = $3 \
             AND tc.status = 'APPROVED'"
        )
        .bind(school_id).bind(class_id).bind(subject_id)
        .fetch_all(&mut *conn)
        .await?;

        if slots.is_empty() {
            return Err(anyhow!("No approved timetable slots found for class {} subject {}", class_id, subject_id));
        }

        // 2. Fetch syllabus_calendar for this class-subject (pending chapters)
        let pending_chapters = sqlx::query(
            "SELECT sc.id, sc.chapter_id, c.name as chapter_name, sc.quarter, sc.period_count \
             FROM syllabus_calendar sc \
             JOIN chapters c ON c.id = sc.chapter_id \
             WHERE sc.school_id = $1 AND sc.class_id = $2 AND sc.subject_id = $3 \
             AND sc.status = 'pending' ORDER BY sc.planned_start_date"
        )
        .bind(school_id).bind(class_id).bind(subject_id)
        .fetch_all(&mut *conn)
        .await?;

        if pending_chapters.is_empty() {
            return Err(anyhow!("No pending syllabus chapters for class {} subject {}", class_id, subject_id));
        }

        // 3. Build a map: day_of_week -> period_numbers for this subject
        let mut day_periods: std::collections::HashMap<i32, Vec<(i32, String, String)>> = std::collections::HashMap::new();
        for row in &slots {
            let dow: i32 = row.get("day_of_week");
            let period: i32 = row.get("period_number");
            let teacher_id: String = row.get("teacher_id");
            let config_id: String = row.get("config_id");
            day_periods.entry(dow).or_default().push((period, teacher_id, config_id));
        }

        // 4. Generate period plans
        let mut plans = Vec::new();
        let mut d = from;
        let mut chapter_idx = 0usize;

        while d <= to {
            let dow = d.format("%u").to_string().parse::<i32>().unwrap_or(0);
            if let Some(periods) = day_periods.get(&dow) {
                for (period_num, teacher_id, config_id) in periods {
                    let sc_row = &pending_chapters[chapter_idx.min(pending_chapters.len() - 1)];
                    let sc_id: i32 = sc_row.get("id");
                    let chapter_id: i32 = sc_row.get("chapter_id");
                    let chapter_name: String = sc_row.get("chapter_name");
                    let topic_name = format!("{} (Week {})", chapter_name, d.iso_week().week());

                    sqlx::query(
                        "INSERT INTO period_plans (school_id, class_id, subject_id, config_id, \
                         day_of_week, period_number, date, chapter_id, topic_name, teacher_id, status) \
                         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, 'pending') \
                         ON CONFLICT (school_id, config_id, day_of_week, period_number, date) \
                         DO UPDATE SET topic_name = EXCLUDED.topic_name, chapter_id = EXCLUDED.chapter_id"
                    )
                    .bind(school_id).bind(class_id).bind(subject_id).bind(config_id)
                    .bind(dow).bind(period_num).bind(d)
                    .bind(chapter_id).bind(&topic_name).bind(teacher_id)
                    .execute(&mut *conn)
                    .await?;

                    plans.push(json!({
                        "date": d.to_string(),
                        "periodNumber": period_num,
                        "chapterId": chapter_id,
                        "topicName": topic_name,
                        "teacherId": teacher_id,
                        "status": "pending",
                    }));

                    chapter_idx = (chapter_idx + 1) % pending_chapters.len();
                }
            }
            d += chrono::Duration::days(1);
        }

        // 5. Mark syllabus_calendar status as in_progress
        if !pending_chapters.is_empty() {
            let first_id: i32 = pending_chapters[0].get("id");
            sqlx::query("UPDATE syllabus_calendar SET status = 'in_progress' WHERE id = $1")
                .bind(first_id).execute(&mut *conn).await?;
        }

        Ok(json!({
            "success": true,
            "classId": class_id,
            "subjectId": subject_id,
            "fromDate": from_date,
            "toDate": to_date,
            "totalPlans": plans.len(),
            "plans": plans,
        }))
    }

    /// Restructure pending period plans when syllabus is behind schedule.
    /// Finds uncompleted plans before `date`, redistributes within the same quarter.
    pub async fn restructure_syllabus_on_delay(
        &self,
        school_id: &str,
        teacher_id: &str,
        date: &str,
    ) -> Result<Value> {
        let mut conn = self.repos.db_client.acquire_tenant_connection(school_id).await?;
        let cutoff = NaiveDate::parse_from_str(date, "%Y-%m-%d")?;

        // 1. Find pending/missed period_plans for this teacher before the cutoff
        let missed = sqlx::query(
            "SELECT pp.* FROM period_plans pp \
             WHERE pp.school_id = $1 AND pp.teacher_id = $2 AND pp.date < $3 \
             AND pp.status IN ('pending', 'missed') \
             ORDER BY pp.date, pp.period_number"
        )
        .bind(school_id).bind(teacher_id).bind(cutoff)
        .fetch_all(&mut *conn)
        .await?;

        if missed.is_empty() {
            return Ok(json!({"success": true, "restructured": 0, "message": "No pending plans found"}));
        }

        // 2. Find future available slots in the same quarter
        let future_slots = sqlx::query(
            "SELECT pp.* FROM period_plans pp \
             WHERE pp.school_id = $1 AND pp.teacher_id = $2 AND pp.date >= $3 \
             AND pp.status = 'pending' \
             ORDER BY pp.date, pp.period_number \
             LIMIT $4"
        )
        .bind(school_id).bind(teacher_id).bind(cutoff)
        .bind(missed.len() as i64)
        .fetch_all(&mut *conn)
        .await?;

        let redistribute_count = missed.len().min(future_slots.len());

        // 3. Redistribute: assign missed topics to future empty slots
        for i in 0..redistribute_count {
            let missed_row = &missed[i];
            let slot_row = &future_slots[i];

            let missed_chapter: Option<i32> = missed_row.get("chapter_id");
            let missed_topic: Option<String> = missed_row.get("topic_name");

            // Update the future slot with the missed topic
            sqlx::query(
                "UPDATE period_plans SET chapter_id = $1, topic_name = $2, status = 'pending' \
                 WHERE school_id = $3 AND id = $4"
            )
            .bind(missed_chapter).bind(&missed_topic)
            .bind(school_id)
            .bind(slot_row.get::<i32, &str>("id"))
            .execute(&mut *conn)
            .await?;

            // Mark the original missed plan as rescheduled
            sqlx::query(
                "UPDATE period_plans SET status = 'rescheduled' WHERE school_id = $1 AND id = $2"
            )
            .bind(school_id)
            .bind(missed_row.get::<i32, &str>("id"))
            .execute(&mut *conn)
            .await?;
        }

        // 4. If some missed topics couldn't be rescheduled → generate alert
        let remaining = missed.len() - redistribute_count;
        if remaining > 0 {
            sqlx::query(
                "INSERT INTO schedule_change_requests (school_id, type, requested_by, reason, status, date_from, date_to) \
                 VALUES ($1, 'skip', $2, $3, 'pending', $4, $5)"
            )
            .bind(school_id).bind(teacher_id)
            .bind(format!("{} topics could not be rescheduled within the quarter — needs admin review", remaining))
            .bind(cutoff).bind(cutoff + chrono::Duration::days(7))
            .execute(&mut *conn)
            .await?;
        }

        Ok(json!({
            "success": true,
            "restructured": redistribute_count,
            "remaining": remaining,
            "message": format!("Rescheduled {} topics. {} still pending admin review.", redistribute_count, remaining),
        }))
    }
}

File: Backend/python/app/services/ai/syllabus_planner.py
from datetime import datetime, date, timedelta
from typing import Dict, List, Any
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

class SyllabusPlanner:
    @staticmethod
    def get_weekday_iso(d: date) -> int:
        # ISO weekday: Monday is 1, Sunday is 7
        return d.isoweekday()

    @staticmethod
    async def annual_syllabus_plot(
        session: AsyncSession,
        school_id: str,
        class_id: str,
        subject_id: str,
        academic_year: int
    ) -> Dict[str, Any]:
        """Plots annual syllabus and distributes chapters across Q1-Q4 based on weightage, ignoring holidays."""
        # 1. Fetch chapters
        result = await session.execute(
            text(
                "SELECT id, name, COALESCE(weightage, 1) as weightage FROM chapters "
                "WHERE school_id = :sid AND subject_id = :sub_id "
                "ORDER BY sequence_order"
            ),
            {"sid": school_id, "sub_id": subject_id}
        )
        chapters = [dict(row._mapping) for row in result.all()]
        if not chapters:
            raise ValueError(f"No chapters found for subject {subject_id}")

        # 2. Fetch holidays in academic year (April 1 to March 31)
        year_start = date(academic_year, 4, 1)
        year_end = date(academic_year + 1, 3, 31)
        
        holiday_result = await session.execute(
            text(
                "SELECT date FROM holidays WHERE school_id = :sid AND date >= :start AND date <= :end"
            ),
            {"sid": school_id, "start": year_start, "end": year_end}
        )
        holidays = {row[0] for row in holiday_result.all()}

        # 3. Define Quarters boundaries
        quarters = [
            ("Q1", date(academic_year, 4, 1), date(academic_year, 6, 30)),
            ("Q2", date(academic_year, 7, 1), date(academic_year, 9, 30)),
            ("Q3", date(academic_year, 10, 1), date(academic_year, 12, 31)),
            ("Q4", date(academic_year + 1, 1, 1), date(academic_year + 1, 3, 31))
        ]

        # 4. Teaching days per quarter (excluding Sundays and holidays)
        quarter_days = {}
        for q_name, q_start, q_end in quarters:
            days = []
            curr = q_start
            while curr <= q_end:
                if curr.isoweekday() != 7 and curr not in holidays:
                    days.append(curr)
                curr += timedelta(days=1)
            quarter_days[q_name] = days

        # 5. Distribute chapters optimally across quarters based on weightage
        total_weight = sum(ch["weightage"] for ch in chapters)
        if total_weight == 0:
            raise ValueError("Total weightage is zero, cannot distribute")

        n = len(chapters)
        target = total_weight / 4.0

        if n < 4:
            p1 = 0
            p2 = min(1, n - 1)
            p3 = min(2, n - 1)
        else:
            best = (0, 1, 2)
            min_err = float("inf")
            for i in range(n):
                for j in range(i + 1, n):
                    for k in range(j + 1, n):
                        if k >= n - 1:
                            continue
                        w1 = sum(ch["weightage"] for ch in chapters[0:i+1])
                        w2 = sum(ch["weightage"] for ch in chapters[i+1:j+1])
                        w3 = sum(ch["weightage"] for ch in chapters[j+1:k+1])
                        w4 = sum(ch["weightage"] for ch in chapters[k+1:])
                        
                        err = (w1 - target)**2 + (w2 - target)**2 + (w3 - target)**2 + (w4 - target)**2
                        if err < min_err:
                            min_err = err
                            best = (i, j, k)
            p1, p2, p3 = best

        # Quarter mappings
        quarter_ranges = [
            (range(0, p1 + 1), "Q1"),
            (range(p1 + 1, p2 + 1), "Q2"),
            (range(p2 + 1, p3 + 1), "Q3"),
            (range(p3 + 1, n), "Q4")
        ]

        inserted = []
        for ch_range, q_name in quarter_ranges:
            q_days = quarter_days[q_name]
            range_size = len(ch_range)
            
            for ch_idx_in_q, ch_idx in enumerate(ch_range):
                ch = chapters[ch_idx]
                chapter_id = ch["id"]
                chapter_name = ch["name"]
                ch_weight = ch["weightage"]
                
                day_start_idx = (ch_idx_in_q * len(q_days)) // max(range_size, 1)
                day_end_idx = ((ch_idx_in_q + 1) * len(q_days)) // max(range_size, 1)
                
                # Fetch start and end date boundaries
                start_d = q_days[min(day_start_idx, len(q_days) - 1)] if q_days else year_start
                end_d = q_days[min(day_end_idx, len(q_days) - 1)] if q_days else year_end
                
                # Period count logic
                period_count = int(round((ch_weight / total_weight) * (len(q_days) * 6.0)))
                
                await session.execute(
                    text(
                        "INSERT INTO syllabus_calendar (school_id, class_id, subject_id, chapter_id, "
                        "planned_start_date, planned_end_date, period_count, quarter, status) "
                        "VALUES (:sid, :cid, :sub, :chid, :pstart, :pend, :pcount, :q, 'pending') "
                        "ON CONFLICT (school_id, class_id, subject_id, chapter_id, quarter) "
                        "DO UPDATE SET planned_start_date = EXCLUDED.planned_start_date, "
                        "planned_end_date = EXCLUDED.planned_end_date, "
                        "period_count = EXCLUDED.period_count"
                    ),
                    {
                        "sid": school_id,
                        "cid": class_id,
                        "sub": subject_id,
                        "chid": chapter_id,
                        "pstart": start_d,
                        "pend": end_d,
                        "pcount": period_count,
                        "q": q_name
                    }
                )
                
                inserted.append({
                    "chapterId": chapter_id,
                    "chapterName": chapter_name,
                    "quarter": q_name,
                    "plannedStartDate": str(start_d),
                    "plannedEndDate": str(end_d),
                    "periodCount": period_count
                })
        
        await session.commit()
        return {
            "success": True,
            "academicYear": f"{academic_year}-{academic_year + 1}",
            "subjectId": subject_id,
            "totalChapters": n,
            "syllabus": inserted
        }

    @staticmethod
    async def micro_plan_period_level(
        session: AsyncSession,
        school_id: str,
        class_id: str,
        subject_id: str,
        from_date_str: str,
        to_date_str: str
    ) -> Dict[str, Any]:
        """Maps class timetable slots to pending chapters to plan day-by-day topics."""
        from_date = datetime.strptime(from_date_str, "%Y-%m-%d").date()
        to_date = datetime.strptime(to_date_str, "%Y-%m-%d").date()

        # 1. Fetch timetable slots
        slots_result = await session.execute(
            text(
                "SELECT ts.day_of_week, ts.period_number, ts.teacher_id, tc.config_id "
                "FROM timetable_slots ts "
                "JOIN timetable_configs tc ON tc.config_id = ts.config_id AND tc.school_id = ts.school_id "
                "WHERE ts.school_id = :sid AND ts.class_id = :cid AND ts.subject_id = :sub "
                "AND tc.status = 'APPROVED'"
            ),
            {"sid": school_id, "cid": class_id, "sub": subject_id}
        )
        slots = slots_result.all()
        if not slots:
            raise ValueError(f"No approved timetable slots found for class {class_id} subject {subject_id}")

        # 2. Fetch pending syllabus chapters
        chapters_result = await session.execute(
            text(
                "SELECT sc.id, sc.chapter_id, c.name as chapter_name, sc.quarter, sc.period_count "
                "FROM syllabus_calendar sc "
                "JOIN chapters c ON c.id = sc.chapter_id "
                "WHERE sc.school_id = :sid AND sc.class_id = :cid AND sc.subject_id = :sub "
                "AND sc.status = 'pending' ORDER BY sc.planned_start_date"
            ),
            {"sid": school_id, "cid": class_id, "sub": subject_id}
        )
        pending_chapters = chapters_result.all()
        if not pending_chapters:
            raise ValueError(f"No pending syllabus chapters for class {class_id} subject {subject_id}")

        # 3. Map day_of_week -> slots
        day_periods = {}
        for row in slots:
            dow, period_num, teacher_id, config_id = row
            day_periods.setdefault(dow, []).append((period_num, teacher_id, config_id))

        # 4. Generate period plans
        plans = []
        curr = from_date
        chapter_idx = 0
        
        while curr <= to_date:
            dow = curr.isoweekday()
            if dow in day_periods:
                for period_num, teacher_id, config_id in day_periods[dow]:
                    sc_row = pending_chapters[chapter_idx % len(pending_chapters)]
                    sc_id, chapter_id, chapter_name, quarter, period_count = sc_row
                    
                    # Compute topic name
                    year, week_num, _ = curr.isocalendar()
                    topic_name = f"{chapter_name} (Week {week_num})"

                    await session.execute(
                        text(
                            "INSERT INTO period_plans (school_id, class_id, subject_id, config_id, "
                            "day_of_week, period_number, date, chapter_id, topic_name, teacher_id, status) "
                            "VALUES (:sid, :cid, :sub, :conf, :dow, :pnum, :date, :chid, :topic, :teacher, 'pending') "
                            "ON CONFLICT (school_id, config_id, day_of_week, period_number, date) "
                            "DO UPDATE SET topic_name = EXCLUDED.topic_name, chapter_id = EXCLUDED.chapter_id"
                        ),
                        {
                            "sid": school_id, "cid": class_id, "sub": subject_id, "conf": config_id,
                            "dow": dow, "pnum": period_num, "date": curr,
                            "chid": chapter_id, "topic": topic_name, "teacher": teacher_id
                        }
                    )
                    
                    plans.append({
                        "date": str(curr),
                        "periodNumber": period_num,
                        "chapterId": chapter_id,
                        "topicName": topic_name,
                        "teacherId": teacher_id,
                        "status": "pending"
                    })
                    
                    chapter_idx += 1
            curr += timedelta(days=1)

        # 5. Mark first pending calendar chapter as in_progress
        if pending_chapters:
            first_sc_id = pending_chapters[0][0]
            await session.execute(
                text("UPDATE syllabus_calendar SET status = 'in_progress' WHERE id = :id"),
                {"id": first_sc_id}
            )

        await session.commit()
        return {
            "success": True,
            "classId": class_id,
            "subjectId": subject_id,
            "fromDate": from_date_str,
            "toDate": to_date_str,
            "totalPlans": len(plans),
            "plans": plans
        }

    @staticmethod
    async def restructure_syllabus_on_delay(
        session: AsyncSession,
        school_id: str,
        teacher_id: str,
        date_str: str
    ) -> Dict[str, Any]:
        """Reschedules missed lessons forward into future empty slots, alerts admin on overflow."""
        cutoff_date = datetime.strptime(date_str, "%Y-%m-%d").date()

        # 1. Fetch pending missed period plans
        missed_result = await session.execute(
            text(
                "SELECT id, chapter_id, topic_name FROM period_plans "
                "WHERE school_id = :sid AND teacher_id = :tid AND date < :cutoff "
                "AND status IN ('pending', 'missed') "
                "ORDER BY date, period_number"
            ),
            {"sid": school_id, "tid": teacher_id, "cutoff": cutoff_date}
        )
        missed = missed_result.all()
        if not missed:
            return {"success": True, "restructured": 0, "message": "No pending plans found"}

        # 2. Fetch future empty slots (limit to the number of missed items)
        future_result = await session.execute(
            text(
                "SELECT id FROM period_plans "
                "WHERE school_id = :sid AND teacher_id = :tid AND date >= :cutoff "
                "AND status = 'pending' "
                "ORDER BY date, period_number "
                "LIMIT :lim"
            ),
            {"sid": school_id, "tid": teacher_id, "cutoff": cutoff_date, "lim": len(missed)}
        )
        future_slots = future_result.all()
        redistribute_count = min(len(missed), len(future_slots))

        # 3. Update future plans with missed chapters/topics
        for i in range(redistribute_count):
            m_row = missed[i]
            f_row = future_slots[i]
            
            # Update future slot
            await session.execute(
                text(
                    "UPDATE period_plans SET chapter_id = :chid, topic_name = :topic, status = 'pending' "
                    "WHERE school_id = :sid AND id = :fid"
                ),
                {
                    "chid": m_row[1],
                    "topic": m_row[2],
                    "sid": school_id,
                    "fid": f_row[0]
                }
            )

            # Mark old missed slot as rescheduled
            await session.execute(
                text("UPDATE period_plans SET status = 'rescheduled' WHERE school_id = :sid AND id = :mid"),
                {"sid": school_id, "mid": m_row[0]}
            )

        # 4. Handle overflow and alerts
        remaining = len(missed) - redistribute_count
        if remaining > 0:
            await session.execute(
                text(
                    "INSERT INTO schedule_change_requests (school_id, type, requested_by, reason, status, date_from, date_to) "
                    "VALUES (:sid, 'skip', :tid, :reason, 'pending', :from_d, :to_d)"
                ),
                {
                    "sid": school_id,
                    "tid": teacher_id,
                    "reason": f"{remaining} topics could not be rescheduled within the quarter — needs admin review",
                    "from_d": cutoff_date,
                    "to_d": cutoff_date + timedelta(days=7)
                }
            )

        await session.commit()
        return {
            "success": True,
            "restructured": redistribute_count,
            "remaining": remaining,
            "message": f"Rescheduled {redistribute_count} topics. {remaining} still pending admin review."
        }
